impl/src/main/java/org/ehcache/internal/classes/ClassInstanceProvider.java

```diff
@@ -45,9 +47,11 @@ protected ClassInstanceProvider(final Class<? extends ClassInstanceProviderFacto
   protected T newInstance(final String alias, final CacheConfiguration<?, ?> cacheConfiguration) {
     Class<? extends T> clazz = null;
     for (ServiceConfiguration<?> serviceConfiguration : cacheConfiguration.getServiceConfigurations()) {
-      if(cacheLevelConfig.isAssignableFrom(serviceConfiguration.getClass())) {
+      if(cacheLevelConfig.isAssignableFrom(serviceConfiguration.getClass()) && !configAdded.contains(serviceConfiguration)) {
```

Comment:
This now works only because the iteration happens in the same order as in `EhcacheManager` which bothers me a lot.
On a side note, I know some have opinions about not supporting more than one listener. So maybe that point should be discussed first. Could you start a conversation for this on the dev mailing list?